src/csharp/Grpc.Core/Internal/AsyncCall.cs

```diff
@@ -78,7 +78,12 @@ public TResponse UnaryCall(TRequest msg)
             var profiler = Profilers.ForCurrentThread();
 
             using (profiler.NewScope("AsyncCall.UnaryCall"))
-            using (CompletionQueueSafeHandle cq = CompletionQueueSafeHandle.CreateSync())
+
+            // Create a pluckable completion queue for the call. Avoid creating a completion queue when we know the channel has already
+            // been shutdown. In such case, the call will fail with ObjectDisposedException immediately anyway and creating / destroying
+            // a completion queue would lead to crash if this was the last channel in the application (and thus GrpcEnvironment has been shutdown).
+            // See https://github.com/grpc/grpc/issues/19090
+            using (CompletionQueueSafeHandle cq = details.Channel.Handle.IsClosed ?  null : CompletionQueueSafeHandle.CreateSync())
```

Comment:
this seems racey, because the channel could still be destroyed from another thread, after creating the cq here, but before the cq destructor runs, right?

Can we fix this instead by having not only the channel but also the completion queue also take a "ref" to the gRPC library? E.g. add a `GrpcEnvironment.GrpcNativeInit();` and `GrpcEnvironment.GrpcNativeShutdown` pair to `CompletionQueueSafeHandle`'s constructor and ReleaseHandle?

Reply:
The problem with completion queue incrementing refcount on GrpcEnvironment is that it requires a global lock. Global lock is fine for creating channels/server because their creation/shutdown is heavyweight and quite rare.
On the other hand, for sync unary calls, a new completion queue is created for each call, so the cost of acquiring a global lock seems prohibitive.

I agree that the fix is not 100% but it would improve the current situation a lot.
I can add a comment that the fix is known to be racy, but it's a big improvement and doesn't have an associated performance hit.

Btw this fix protects against a situation that should not happen at all under normal circumstances (no point of starting a call after you've shutdown your channel). It's basically a protection against users shooting themselves in the foot when the write misbehaving code.

WDYT?

Reply:
I could see how there would be performance benefits to not acquiring the lock, but my concerns are:
a) I have doubts that the magnitude of the gain is worth it - this lock should be uncontended in the common case, unless people are doing lots of concurrent sync RPCs from different threads (in which case they should probably be using async API).
b) Even though this should improve things a lot, the potential crash will be hard to debug if hit.
c) Creating a new CQ, reffing the grpc library, and taking a global lock while doing so, are I believe exactly what the C++ sync API does (when it calls grpc_init and grpc_shutdown in the ctor and dtor of it's CQ object).

Reply:
The race condition was an issue even without this change. The old line: ```using (CompletionQueueSafeHandle cq = CompletionQueueSafeHandle.CreateSync())``` could still complete and then have the channel deleted by another thread before the destructor is called. 
Since this also doesn't add any new races and resolves the user's issue, I think it's best to push this change in and address the race separately as it isn't new. 
I can't see any other way to fix the race without using some form of a global lock. Since the plan is to deprecate this library, creating the PR and monitoring for performance is probably more effort than it's worth.